enterprise/server/usage/usage.go

```diff
@@ -0,0 +1,151 @@
+package usage
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/tables"
+	"github.com/buildbuddy-io/buildbuddy/server/util/perms"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/buildbuddy-io/buildbuddy/server/util/timeutil"
+	"github.com/go-redis/redis"
+	"github.com/jonboulle/clockwork"
+)
+
+const (
+	redisUsageKeyPrefix       = "usage/"
+	redisCountsKeyPrefix      = redisUsageKeyPrefix + "counts/"
+	redisInvocationsKeyPrefix = redisUsageKeyPrefix + "invocations/"
+)
+
+type TrackerOpts struct {
+	Clock clockwork.Clock
+}
+
+type tracker struct {
+	env   environment.Env
+	rdb   *redis.Client
+	clock clockwork.Clock
+}
+
+func NewTracker(env environment.Env) (*tracker, error) {
+	return NewTrackerWithOpts(env, &TrackerOpts{})
+}
+
+func NewTrackerWithOpts(env environment.Env, opts *TrackerOpts) (*tracker, error) {
+	rdb := env.GetUsageRedisClient()
+	if rdb == nil {
+		return nil, status.UnimplementedError("Missing redis client for usage")
+	}
+	clock := opts.Clock
+	if clock == nil {
+		clock = clockwork.NewRealClock()
+	}
+	return &tracker{
+		env:   env,
+		rdb:   rdb,
+		clock: clock,
+	}, nil
+}
+
+func (ut *tracker) Increment(ctx context.Context, counts *tables.UsageCounts) error {
+	groupID, err := authenticatedGroupID(ctx, ut.env)
+	if err != nil {
+		return err
+	}
+	if groupID == "" {
+		// Don't track anonymous usage for now.
+		return nil
+	}
+
+	m, err := countsToMap(counts)
+	if err != nil {
+		return err
+	}
+	if len(m) == 0 {
+		return nil
+	}
+
+	uk := countsRedisKey(groupID, ut.clock.Now())
+	pipe := ut.rdb.TxPipeline()
+	for k, c := range m {
+		pipe.HIncrBy(ctx, uk, k, c)
```

Comment:
I think we should expire these keys after 2-3x the flush period. It limits what we have to potentially scan (which is nice), and keeps us from potentially polluting redis with values we never read.

If we haven't picked them up in a couple of flush periods, we're not going to. (Also, in that case, would we even be logging new data? Seems like something bigger is wrong)